rmw_opendds_cpp/src/rmw_service.cpp

```diff
@@ -26,319 +26,241 @@
+        info->callbacks_->get_reply_datawriter(info->replier_));
+      if (reply_datawriter) {
+        node_info.publisher_listener->remove_information(
+          reply_datawriter->get_instance_handle(), EntityType::Publisher);
+        node_info.publisher_listener->trigger_graph_guard_condition();
+      }
+      info->callbacks_->destroy_replier(info->replier_, &rmw_free);
+      info->replier_ = nullptr;
+    }
+
+    RMW_TRY_DESTRUCTOR(info->~OpenDDSStaticServiceInfo(), OpenDDSStaticServiceInfo, ret = RMW_RET_ERROR)
+    rmw_free(info);
+    service->data = nullptr;
+  }
+
+  if (service->service_name) {
+    rmw_free(const_cast<char *>(service->service_name));
+    service->service_name = nullptr;
+  }
+  rmw_service_free(service);
+  return ret;
+}
+
 rmw_service_t *
 rmw_create_service(
   const rmw_node_t * node,
   const rosidl_service_type_support_t * type_supports,
   const char * service_name,
   const rmw_qos_profile_t * qos_profile)
 {
-  if (!node) {
-    RMW_SET_ERROR_MSG("node is null");
-    return NULL;
-  }
-  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
-    node handle, node->implementation_identifier, opendds_identifier,
-    return NULL)
+  RMW_CHECK_FOR_NULL_WITH_MSG(node, "node is null", return nullptr);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(node, node->implementation_identifier, opendds_identifier, return nullptr)
+
+  auto node_info = static_cast<OpenDDSNodeInfo *>(node->data);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info, "node_info is null", return nullptr);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info->participant.in(), "participant is null", return nullptr);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info->publisher_listener, "publisher_listener is null", return nullptr);
+  RMW_CHECK_FOR_NULL_WITH_MSG(node_info->subscriber_listener, "subscriber_listener is null", return nullptr);
 
   const rosidl_service_type_support_t * type_support = rmw_get_service_type_support(type_supports);
   if (!type_support) {
-    return NULL;
-  }
-
-  if (!qos_profile) {
-    RMW_SET_ERROR_MSG("qos_profile is null");
     return nullptr;
   }
 
-  auto node_info = static_cast<OpenDDSNodeInfo *>(node->data);
-  if (!node_info) {
-    RMW_SET_ERROR_MSG("node info is null");
-    return NULL;
-  }
-  auto participant = static_cast<DDS::DomainParticipant *>(node_info->participant);
-  if (!participant) {
-    RMW_SET_ERROR_MSG("participant is null");
-    return NULL;
-  }
-
-  const service_type_support_callbacks_t * callbacks =
-    static_cast<const service_type_support_callbacks_t *>(type_support->data);
-  if (!callbacks) {
-    RMW_SET_ERROR_MSG("callbacks handle is null");
-    return NULL;
+  RMW_CHECK_FOR_NULL_WITH_MSG(service_name, "service_name is null", return nullptr);
+  CORBA::String_var request_topic;
+  CORBA::String_var response_topic;
+  if (!_process_service_name(service_name, qos_profile->avoid_ros_namespace_conventions,
+      &request_topic.out(), &response_topic.out())) {
```

Comment:
I'm not 100% sure this use of `.out` is going to work.  Since `_process_service_name` is internal we could change it's argument types to be friendlier (for example, use std::string).

Reply:
used std::string